### PR TITLE
Porting SOF vBE

### PR DESCRIPTION
**Pull request by libinyang**
The first patch is from Yingjiang and Please ignore it. Thanks.

### CODE REVIEW

File: sound/soc/sof/core.c [hunk 1]
```diff
@@ -314,22 +326,18 @@ static int sof_probe(struct platform_device *pdev)
...
 	/* init DMA trace */
 	ret = snd_sof_init_trace(sdev);
 	if (ret < 0) {
 		/* non fatal */
 		dev_warn(sdev->dev,
 			 "warning: failed to initialize trace %d\n", ret);
 	}
-
-	/* autosuspend sof device */
-	pm_runtime_mark_last_busy(sdev->dev);
```
Comment (lgirdwood) [inline sound/soc/sof/core.c:326]: We still need PM for BE and non virtio mode.

File: sound/soc/sof/sof-priv.h [hunk 2]
```diff
@@ -22,9 +22,29 @@
 #include <uapi/sound/sof-ipc.h>
 #include <uapi/sound/sof-fw.h>
 #include <uapi/sound/asoc.h>
+#include <uapi/sound/sof-virtio.h>
 #include <sound/hdaudio.h>
 #include <sound/compress_driver.h>
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_VIRTIO_FE)
+// TODO refine which ones are needed
```
Comment (lgirdwood) [inline sound/soc/sof/sof-priv.h:30]: Has refinement of headers been done, if so we can remove.

File: sound/soc/sof/sof-priv.h [hunk 3]
```diff
@@ -283,6 +308,79 @@ struct snd_sof_dai {
...
+#define SOF_VIRTIO_IPC_REPLY 1
+struct sof_vbe_client {
+	struct sof_vbe *vbe;
+	int vhm_client_id;
+	int max_vcpu;
+	struct vhm_request *req_buf;
+	struct list_head list;
+
+	// TODO: allow clients to have N streams. Map streams to posn[n]
+	struct sof_ipc_stream_posn pos[4];
```
Comment (lgirdwood) [inline sound/soc/sof/sof-priv.h:327]: Best to define a macro here for 4.

File: sound/soc/sof/sof-priv.h [hunk 4]
```diff
@@ -362,6 +471,18 @@ struct snd_sof_dev {
 	void *private;			/* core does not touch this */
 };
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_VIRTIO_BE)
+int snd_sof_virtio_miscdev_register(struct snd_sof_dev *sdev);
+int snd_sof_virtio_miscdev_unregister(void);
+#else
+static int snd_sof_virtio_miscdev_register(struct snd_sof_dev *sdev)
```
Comment (lgirdwood) [inline sound/soc/sof/sof-priv.h:485]: static inline

File: sound/soc/sof/sof-priv.h [hunk 5]
```diff
@@ -477,10 +598,27 @@ int snd_sof_get_status(struct snd_sof_dev *sdev, u32 panic_code,
 		       size_t stack_size);
 int snd_sof_init_trace_ipc(struct snd_sof_dev *sdev);
 
+/*
+ * VirtIO
+ */
+#define GUEST_COMP_ID_OFFSET  0x100
+int sof_virtio_submit_guest_ipc(struct snd_sof_dev *sdev, int vm_id,
```
Comment (lgirdwood) [inline sound/soc/sof/sof-priv.h:612]: if these are private APIs to only be used within SOF then we can prefix them as sof_virtio_

File: sound/soc/sof/virtio-fe.c [hunk 6]
```diff
@@ -19,112 +20,426 @@
...
+	};
+
+	dev = &vdev->dev;
 
-	/* conenct virt queues to BE */
+	vfe = devm_kzalloc(dev, sizeof(*vfe), GFP_KERNEL);
+	if (!vfe)
+		return -ENOMEM;
+
+	//sdev->vfe = vfe;
```
Comment (lgirdwood) [inline sound/soc/sof/virtio-fe.c:342]: needed now ?

File: sound/soc/sof/virtio-fe.c [hunk 7]
```diff
@@ -19,112 +20,426 @@
...
+static int vfe_probe(struct virtio_device *vdev)
 {
-	/* register virtio device */
+	struct virtqueue *vqs[SOF_VIRTIO_NUM_OF_VQS];
+	struct device *dev;
+	struct scatterlist sg;
+	struct sof_vfe *vfe;
+	int ret;
+
+	/* the processing callback number must be the same as the vqueues.*/
```
Comment (lgirdwood) [inline sound/soc/sof/virtio-fe.c:328]: This should probably be declared as a static array ?

File: sound/soc/sof/virtio-fe.c [hunk 8]
```diff
@@ -19,112 +20,426 @@
...
+static void vfe_remove(struct virtio_device *vdev)
 {
 	/* free virtio resurces and unregister device */
+	struct sof_vfe *vfe = vdev->priv;
+
+	vdev->config->reset(vdev);
+	vdev->config->del_vqs(vdev);
+	kfree(vfe->posn);
+
+	// unregister the devices of SOF
```
Comment (lgirdwood) [inline sound/soc/sof/virtio-fe.c:387]: C style comments

File: sound/soc/sof/virtio-miscdev.c [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+ */
+
+static struct virtio_miscdev *virtio_audio;
+
+static struct virtio_miscdev *get_virtio_audio(void)
+{
+	return virtio_audio;
+}
+
+void *get_sof_dev(void)
```
Comment (lgirdwood) [inline sound/soc/sof/virtio-miscdev.c:38]: prefix with sof_virtio_

File: sound/soc/sof/virtio-miscdev.c [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+		return ret;
+
+	vaudio->open = sof_virtio_open;
+	vaudio->ioctl = sof_virtio_ioctl;
+	vaudio->release = sof_virtio_release;
+
+	return 0;
+}
+
+int snd_sof_virtio_miscdev_unregister(void)
```
Comment (lgirdwood) [inline sound/soc/sof/virtio-miscdev.c:136]: needed ?

File: sound/soc/sof/virtio-miscdev.c [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+ * @dev: the device the audio virtio belongs to
+ * @data: the priv data of virtio_miscdev
+ * @va: allocated virtio_miscdev, return to caller
+ *
+ * This function registers the misc device, which will be used
+ * by the user space to communicate with the audio driver.
+ *
+ * Return: 0 for success or negative value for err
+ */
+int snd_audio_virtio_miscdev_register(struct device *dev, void *data,
```
Comment (lgirdwood) [inline sound/soc/sof/virtio-miscdev.c:210]: It looks like we are only calling this internally ? if so it should be static.

File: sound/soc/sof/virtio-be.c [hunk 12]
```diff
@@ -79,10 +85,328 @@ static int sbe_ipc_comp(struct snd_sof_dev *sdev, int vm_id,
...
+		}
+	}
+
+	return NULL;
+}
+
+static int sbe_pcm_open(struct snd_sof_dev *sdev,
+			void *ipc_data, int vm_id)
+{
+	/*
```
Comment (lgirdwood) [inline sound/soc/sof/virtio-be.c]: Stick this comment before func and comment each non obvious block in every function. i.e. I need to know what each block does.

File: sound/soc/sof/core.c (hunk 1 above)
Comment (libinyang) [reply sound/soc/sof/core.c:326]: The first patch is Yingjiang's patch. I use this as a code base. Please @zhuyingjiang help answer.

File: sound/soc/sof/virtio-miscdev.c (hunk 9 above)
Comment (libinyang) [reply sound/soc/sof/virtio-miscdev.c:38]: OK

File: sound/soc/sof/virtio-miscdev.c (hunk 10 above)
Comment (libinyang) [reply sound/soc/sof/virtio-miscdev.c:136]: Do you mean snd_sof_virtio_miscdev_unregister? For module mode, I think yes, we need it when unload the sof modules.

File: sound/soc/sof/virtio-miscdev.c (hunk 11 above)
Comment (libinyang) [reply sound/soc/sof/virtio-miscdev.c:210]: Right. I will change it.

**Comment by libinyang** [comment]
Hi Liam, my first patch is copying from Yingjiang. I think he will submit the patch himself. I push it here because I use it as my code base. Otherwise, our patches will conflict.

**Pull request closed without merging**